Add context argument to registry.select()

coderkevin opened the pull request:
## Description
This adds an optional context argument to `registry.select()` and then adds support for `withSelect` to pass the component as a context for a select operation.

## Why is this needed?
This makes it possible to map select calls to specific components, which means a data system can have knowledge of which components depend on which data. This gives better options for shared data handling that can be tied to the lifecycle of a given component.

## How has this been tested?
All existing tests still pass, and additional tests have been added to ensure the operation of this new, optional feature.

## Types of changes
Adds an optional argument to `register.select( reducerKey, context)`.
Adds code to `withSelect` to supply a context of `{ component }`.

Both changes are fully backwards-compatible.

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/accessibility-coding-standards/ -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/javascript/ -->

coderkevin committed:
data: Add optional context to registry.select()

This adds an optional `context` as a second argument to
`registry.select()` and also the generic store's `getSelectors()`
function. This enables context-mapped selectors, which helps with
declarative data requirements based on things like UI components when
they become mounted and unmounted.

coderkevin committed:
data: Add context to select within withSelect

This adds a context of { component: this } to withSelect's call to
registry.select. It does this by wrapping registry.select with an
anonymous function that fills in the context when it's called.

tofumatt reviewed (commented):
Looks like this requires a rebase to fix the merge conflicts, but I'm also curious why this is needed more specifically? What issue will making the change solve? Is there a particular use case/example you could provide that would use this code? I think it would help to have some extra context here. 😄

coderkevin commented:
Rebased on current master.

@tofumatt I'm trying to integrate another more advanced data system, [fresh-data](https://github.com/Automattic/fresh-data) as a generic data store within `@wordpress/data`. That data system maps API data to the components which need it, which allows it to update its data graph as components mount and unmount, which is really helpful when calculating things like freshness requirements for automatic re-fetching, but this will also be used to clear out API data that was once fetched but no longer needed by any currently mounted component.

With this change, the differences between these systems become transparent to the application developer.

tofumatt commented on packages/data/src/components/with-select/test/index.js:
```diff
@@ -64,6 +64,39 @@ describe( 'withSelect', () => {
+
+		expect( getSelectors ).toHaveBeenCalledTimes( 1 );
+
+		const selectorContext = getSelectors.mock.calls[ 0 ][ 0 ];
```
Just out of curiosity, are there any nicer test matchers/APIs for this? Not a big deal if no, but just curious 😄

coderkevin replied:
It doesn't appear so, this is from the Jest docs:

> // The first arg of the last call to the mock function was `42`
// (note that there is no sugar helper for this specific of an assertion)
expect(mockFunc.mock.calls[mockFunc.mock.calls.length - 1][0]).toBe(42);

tofumatt commented on packages/data/src/components/with-select/test/index.js:
```diff
@@ -64,6 +64,39 @@ describe( 'withSelect', () => {
+	it( 'passes through the component context to the registry store', () => {
+		const getSelectors = jest.fn();
+
+		const genericStore = {
```
I might call this `customStore` or something just to show the intent of this API better. If you don't provide a context we'd fall back to the _actual_ `wp-data` generic store, right?

coderkevin replied:
Updated in f77e80c

tofumatt commented on packages/data/src/test/registry.js:
```diff
@@ -440,6 +440,22 @@ describe( 'createRegistry', () => {
+
+		it( 'passes through optional context to getSelectors', () => {
+			const getSelectors = jest.fn();
+			const genericStore = {
```
Same comment here re: `customStore` or something like that.

coderkevin replied:
Updated in f77e80c

tofumatt reviewed (changes requested):
Seems to work for me and I just have a few little comments, but I think this requires documentation updates before merging.

CC'ing @youknowriad and @aduth; not sure this is okay to merge at this point in the API freeze times.

coderkevin commented:
Rebased onto `origin/master`

coderkevin committed:
data: Add doc for getSelectors( context )

This adds documentation in the wp.data readme for the use of context.
It also renames `genericStore` to `customStore` in the tests.

coderkevin commented:
@tofumatt (and others), I've added docs and updated as requested.

tofumatt reviewed (commented):
Really sorry this took ages to get to—it got buried in my review queue. But the changes look good; if you can rebase onto master then I think this is good to go?

aduth commented on packages/data/README.md:
````diff
@@ -374,6 +386,7 @@ function createCustomStore() {
 registry.registerGenericStore( 'custom-data', createCustomStore() );
 ```
 
+Note in the above example the handling of the optional `context` argument to `getSelectors()`. When using `withSelect`, this provides information about which component is requesting which data from the store. The benefit of this is the ability to keep track of which data is still currently needed by currently mounted components and which data is no longer needed.
````
>The benefit of this is the ability to keep track of [...] which data is no longer needed

How would this be achieved? In the above example, the cache of `componentResources` will grow infinitely (which in its own right is very problematic). A component will not be removed if it's unmounted, nor is it obvious how this would be accomplished.

coderkevin replied:
The implementation I have for wc-admin removes components from the map upon unmount. The only components in the map are ones currently mounted at any given point in time.

coderkevin commented:
Hey there, I just wanted to let you know that I've got this working with a separate [`with-select` component in our repo](https://github.com/woocommerce/wc-admin/blob/master/client/wc-api/with-select.js). I'll reply again here later to discuss more details on how to make this work in a more general way as intended originally.

gziolo commented:
@coderkevin does it mean this PR can be closed or do you plan to refresh it based on the revised solution you figured out?

By the way, we are setting up some more detailed structure for reviewing PRs. Would you like to get automated pings for review for the code inside `packages/data` folder given your very deep understanding of this area? Related PR: #13604 and parent issue #13441.

coderkevin commented:
> does it mean this PR can be closed or do you plan to refresh it based on the revised solution you figured out?

I'm not sure yet. I haven't yet circled back to this PR but hope to in the next week.

> Would you like to get automated pings for review for the code inside packages/data folder given your very deep understanding of this area?

Sure, I'd be happy to help review anything in packages/data. Thanks!

timmyc commented:
@coderkevin are we still going to try and merge this in at some point?

gziolo commented:
This PR would have to cover also newly introduced hooks: `useSelect` and `useDispatch`. I'm marking this PR as stale to reflect that it's fairly outdated.

coderkevin commented:
@timmyc Thanks for the reminder. I've been working on a modification to fresh-data that would no longer require the special handling needed in this PR. So hopefully this PR won't be needed after the next major version of fresh-data.

nerrad commented:
> Thanks for the reminder. I've been working on a modification to fresh-data that would no longer require the special handling needed in this PR. So hopefully this PR won't be needed after the next major version of fresh-data.

How is this coming @coderkevin?  Will this PR (or some variation of it, considering the staleness) be needed?

nerrad commented:
Things have changed in the project inspiring this original pull, so it can be closed.
